### ISSUE TITLE
Limit order bar matching 

### ISSUE THREAD

**Issue by OnlyC** [nautechsystems/nautilus_trader#1258]
# Bug Report
I don't know if this issue is addressed in the Improved limit order backtesting or not. 
Let say `on_bar` we got `bar close > bar low`, we place a limit buy order at the low of the bar. 
Currently, it will be filled right away, like a market order.

### Expected Behavior
It can be filled at later bars or no filled at all.

- `nautilus_trader` version: 1.178

**Comment by cjdsellers** [comment]
Hi @OnlyC

Looking through the code paths when the `OrderMatchingEngine` is processing a bar, I can't see how we would arrive at a situation where the market would be left at the low and not the close https://github.com/nautechsystems/nautilus_trader/blob/develop/nautilus_trader/backtest/matching_engine.pyx#L458.

If you follow each of those, you'll see that the market is left at the bar `close` after each bar is processed.

Would you be able to provide an MRE of the behavior, demonstrating where it's incorrect?

Many thanks!

**Comment by OnlyC** [comment]
Thanks for suggesting! I figured it out.
My catalog includes 4H, 1H, 15m, 5m candles, although I subscribed to only 4H candles, others candles still processed by OrderMatchingEngine. 4H => 1H => 15m => 5m.

![image](https://github.com/nautechsystems/nautilus_trader/assets/1250580/eac20f68-50be-4102-bf09-7300e910f8c8)

I though we can fix it by process short timeframe first: 5m => 15m => 1H => 4H. But this still has caveat in the case: 5m candle can be a child of the next processed 15m candle, if price going up from the first 5m and never go down, it should not fill the limit order on 5m bar.low but it still fill the order when processing the 15m candle low. (bar.low's tick processed twice)

Other way is to add a time filter to OrderMatchingEngine (> ts_event). We solve the order matching but the book still processed multiple times.
![image](https://github.com/nautechsystems/nautilus_trader/assets/1250580/34991c6f-201d-43ad-9ddf-ceae47157533)
